src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1147,14 +1147,16 @@ private Task WriteStringAsync(string s)
             if (s.Length <= _writeBuffer.Length - offset)
             {
                 byte[] writeBuffer = _writeBuffer;
+                int mask = StaticHttpSettings.EncodingValidationMask;
                 foreach (char c in s)
                 {
-                    if ((c & 0xFF80) != 0)
+                    if ((c & mask) != 0)
                     {
                         throw new HttpRequestException(SR.net_http_request_invalid_char_encoding);
                     }
```

Comment:
I wanted to keep this change as minimal as possible, and did not change the exception message `Request headers must contain only ASCII characters.`, neither bothered to introduce a new one for the Latin1 case.

The message might be confusing when `AllowNonAsciiHeaders == true`, but only a minimal amount of users are effected, therefore it's probably not worth the additional complexity and localization effort.

Reply:
You could parametrize the message:
`Request headers must contain only {0} characters.` and pass ASCII or LATIN1

Reply:
Aren't those messages localized? (Which may lead to issues unless the localization is updated.)

Reply:
To answer the question - no we do not localize any resources in netcoreapp. It's fine to change them at any time. If we didn't want to create new messages, we would probably normally extend the existing one so at least it wasn't definitely wrong. But not suggesting you change it now.

Reply:
In the light of this information, are you still against changing the exception message? @antonfirsov

Reply:
Honestly, not sure if it's worth the efforts to change this in the 3.1 branch (-> rerunning all the tests etc). We expect very few (maybe only 1) customer to utilize this feature.

Reply:
Note: we decided offline to not change this.